lib/events.js

```diff
@@ -352,13 +353,32 @@ EventEmitter.prototype.emit = function emit(type, ...args) {
   return true;
 };
 
-function _addListener(target, type, listener, prepend) {
+function _addListener(target, type, listener, prepend, options) {
   let m;
   let events;
   let existing;
 
   checkListener(listener);
 
+  if (options !== undefined) {
+    const signal = options ? options.signal : undefined;
+    validateAbortSignal(signal, 'options.signal');
+    if (signal.aborted) {
+      return target;
+    }
+    // The same listener can be added/removed multiple times
+    // the signal needs to remove the correct instance - so we wrap it
+    const originalListener = listener;
+    listener = function listener(...args) {
+      return originalListener.apply(this, args);
+    };
```

Comment:
Does this play well with `.rawListeners()` and `.listeners()`? I think we’re basically running into the same issue we had with `.once()` listeners here

Reply:
It's a good question - I think what I'm trying to solve is "what happens if someone is adding the same listener several times, some of them with an AbortController and then removing the (middle, for example) listener.

I guess now that you say it that it's exactly the same issue that `once` has - would it make sense to use the same machinary?

I don't see how `once` actually solves this though - should I just _not_ support `removeListener/off` together with signals?

Reply:
Also worth noting, the static versions of `once` and `on` don't have this problem since they do not accept listener arguments so it's impossible to pass the same reference twice.

Reply:
Oh, I think that I misunderstood the code and that Node already pays the penalty by checking `.listener` _anyway_ in order for `once` listeners to work, interesting.

I'll just that machinery and see what the benchmarks look like.

Reply:
Basically, what I think should be the case (and what we should test) is that:

- `.listeners()` returns the original listeners that were passed in, not the wrapped ones
- `.rawListeners()` returns the fully wrapped listeners, so that code like https://github.com/nodejs/node/blob/0ddd69e2eddc415105d099f614c676efe1f886c3/lib/vm.js#L264-L280 keeps working

Reply:
Switching this to the `listener` machinery appeared to make this work ^_^

Reply:
Added a test for that as well, next I'll run the benchmarks on CI too. If there is no measurable regression I'll mark the PR ready though there is one extra check now so there might be a small regression.